### ISSUE TITLE
[TD3强化学习执行策略] | TD3 Execution Strategy

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4541]
**标签**: quantitative-strategy
### 摘要
Twin Delayed DDPG (TD3) 算法用于优化订单执行路径，最小化市场冲击和机会成本。
### 算法描述
**核心原理**：

TD3（Twin Delayed Deep Deterministic Policy Gradient）是一种强化学习算法，用于解决连续动作空间的决策问题。在订单执行场景中，TD3 学习如何将大订单拆分为小订单，在最小化市场冲击的同时控制执行风险。

**数学模型**：

1. **最优执行问题**：
   - 价值函数：$V(s) = \max_{\pi} \mathbb{E}\left[\sum_{t=0}^{T} \gamma^t r(s_t, a_t)\right]$
   - Q函数：$Q(s, a) = \mathbb{E}\left[r(s, a) + \gamma V(s')\right]$
   - 最优Q函数：$Q^*(s, a) = \mathbb{E}\left[r(s, a) + \gamma \max_{a'} Q^*(s', a')\right]$

2. **Twin Critic架构**：
   - 双Q网络：$Q_1(s, a)$ 和 $Q_2(s, a)$，取最小值防止过估计
   - 目标Q值：$y = r + \gamma \min(Q_1(s', a'), Q_2(s', a'))$
   - 延迟策略更新：每2次更新一次策略网络

3. **高斯噪声探索**：
   - 动作噪声：$a_t = \pi(s_t) + \epsilon$，$\epsilon \sim \mathcal{N}(0, \sigma^2)$
   - 目标策略平滑：$a' = \pi(s') + \text{clip}(\epsilon, -c, c)$

4. **经验回放**：
   - 缓冲区：$\mathcal{D} = \{(s, a, r, s', d)\}$
   - 批量采样：随机抽取 mini-batch 进行梯度下降

5. **订单执行成本模型**：
   - 市场冲击：$MI = \eta \cdot \sqrt{\frac{Q}{V}}$（平方根法则）
   - 机会成本：$OC = \mathbb{E}\left[(P_{exec} - P_{arrival}) \cdot \mathbf{1}_{exec < Q}\right]$
   - 总成本：$TC = MI + \lambda \cdot OC$

**策略流程**：
```
1. 初始化Actor-Critic网络和目标网络
2. 收集市场数据（价格、成交量、波动率）
3. 将状态定义为 (time_remaining, shares_remaining, market_features)
4. Actor网络输出每时刻的最优下单量
5. Critic网络评估当前策略的价值
6. 定期更新目标网络和策略网络参数
```

### Python代码示例
```python
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset
from collections import deque
import random


class ReplayBuffer:
    """经验回放缓冲区"""
    
    def __init__(self, capacity: int = 100000):
        self.buffer = deque(maxlen=capacity)
    
    def push(self, state, action, reward, next_state, done):
        self.buffer.append((state, action, reward, next_state, done))
    
    def sample(self, batch_size: int):
        batch = random.sample(self.buffer, min(batch_size, len(self.buffer)))
        states, actions, rewards, next_states, dones = zip(*batch)
        
        return (
            np.array(states, dtype=np.float32),
            np.array(actions, dtype=np.float32),
            np.array(rewards, dtype=np.float32).reshape(-1, 1),
            np.array(next_states, dtype=np.float32),
            np.array(dones, dtype=np.float32).reshape(-1, 1)
        )
    
    def __len__(self):
        return len(self.buffer)


class Actor(nn.Module):
    """策略网络"""
    
    def __init__(self, state_dim: int, action_dim: int, hidden_dim: int = 256):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim // 2),
            nn.ReLU(),
            nn.Linear(hidden_dim // 2, action_dim),
            nn.Tanh()
        )
    
    def forward(self, state):
        return self.net(state) * 2.0


class Critic(nn.Module):
    """Q网络（双critic之一）"""
    
    def __init__(self, state_dim: int, action_dim: int, hidden_dim: int = 256):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(state_dim + action_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim // 2),
            nn.ReLU(),
            nn.Linear(hidden_dim // 2, 1)
        )
    
    def forward(self, state, action):
        x = torch.cat([state, action], dim=1)
        return self.net(x)


class TD3Agent:
    """TD3强化学习智能体"""
    
    def __init__(
        self,
        state_dim: int,
        action_dim: int,
        hidden_dim: int = 256,
        lr: float = 3e-4,
        gamma: float = 0.99,
        tau: float = 0.005,
        policy_noise: float = 0.2,
        noise_clip: float = 0.5,
        policy_freq: int = 2
    ):
        self.state_dim = state_dim
        self.action_dim = action_dim
        self.gamma = gamma
        self.tau = tau
        self.policy_noise = policy_noise
        self.noise_clip = noise_clip
        self.policy_freq = policy_freq
        self.total_it = 0
        
        self.actor = Actor(state_dim, action_dim, hidden_dim).cuda()
        self.actor_target = Actor(state_dim, action_dim, hidden_dim).cuda()
        self.actor_target.load_state_dict(self.actor.state_dict())
        self.actor_optimizer = optim.Adam(self.actor.parameters(), lr=lr)
        
        self.critic1 = Critic(state_dim, action_dim, hidden_dim).cuda()
        self.critic2 = Critic(state_dim, action_dim, hidden_dim).cuda()
        self.critic1_target = Critic(state_dim, action_dim, hidden_dim).cuda()
        self.critic2_target = Critic(state_dim, action_dim, hidden_dim).cuda()
        self.critic1_target.load_state_dict(self.critic1.state_dict())
        self.critic2_target.load_state_dict(self.critic2.state_dict())
        self.critic_optimizer = optim.Adam(
            list(self.critic1.parameters()) + list(self.critic2.parameters()),
            lr=lr
        )
        
        self.replay_buffer = ReplayBuffer()
    
    def select_action(self, state, noise_scale: float = 0.1):
        state_tensor = torch.FloatTensor(state).unsqueeze(0).cuda()
        with torch.no_grad():
            action = self.actor(state_tensor).cpu().numpy()[0]
        
        if noise_scale > 0:
            noise = np.random.normal(0, noise_scale, size=self.action_dim)
            action = np.clip(action + noise, -2, 2)
        
        return action
    
    def train_step(self, batch_size: int = 256):
        if len(self.replay_buffer) < batch_size:
            return
        
        states, actions, rewards, next_states, dones = self.replay_buffer.sample(batch_size)
        
        states_t = torch.FloatTensor(states).cuda()
        actions_t = torch.FloatTensor(actions).cuda()
        rewards_t = torch.FloatTensor(rewards).cuda()
        next_states_t = torch.FloatTensor(next_states).cuda()
        dones_t = torch.FloatTensor(dones).cuda()
        
        with torch.no_grad():
            noise = torch.clamp(
                torch.randn_like(actions_t) * self.policy_noise,
                -self.noise_clip, self.noise_clip
            )
            next_actions = torch.clamp(
                self.actor_target(next_states_t) + noise,
                -2, 2
            )
            
            target_q1 = self.critic1_target(next_states_t, next_actions)
            target_q2 = self.critic2_target(next_states_t, next_actions)
            target_q = torch.min(target_q1, target_q2)
            target_q = rewards_t + self.gamma * (1 - dones_t) * target_q
        
        current_q1 = self.critic1(states_t, actions_t)
        current_q2 = self.critic2(states_t, actions_t)
        
        critic1_loss = nn.MSELoss()(current_q1, target_q)
        critic2_loss = nn.MSELoss()(current_q2, target_q)
        critic_loss = critic1_loss + critic2_loss
        
        self.critic_optimizer.zero_grad()
        critic_loss.backward()
        self.critic_optimizer.step()
        
        self.total_it += 1
        
        if self.total_it % self.policy_freq == 0:
            actor_loss = -self.critic1(states_t, self.actor(states_t)).mean()
            
            self.actor_optimizer.zero_grad()
            actor_loss.backward()
            self.actor_optimizer.step()
            
            for param, target_param in zip(
                self.actor.parameters(), self.actor_target.parameters()
            ):
                target_param.data.copy_(
                    self.tau * param.data + (1 - self.tau) * target_param.data
                )
            
            for param, target_param in zip(
                self.critic1.parameters(), self.critic1_target.parameters()
            ):
                target_param.data.copy_(
                    self.tau * param.data + (1 - self.tau) * target_param.data
                )
            
            for param, target_param in zip(
                self.critic2.parameters(), self.critic2_target.parameters()
            ):
                target_param.data.copy_(
                    self.tau * param.data + (1 - self.tau) * target_param.data
                )


class OrderExecutionSimulator:
    """订单执行模拟器"""
    
    def __init__(
        self,
        initial_price: float = 100.0,
        volatility: float = 0.02,
        mean_reversion: float = 0.1,
        total_shares: int = 100000,
        num_steps: int = 100
    ):
        self.initial_price = initial_price
        self.volatility = volatility
        self.mean_reversion = mean_reversion
        self.total_shares = total_shares
        self.num_steps = num_steps
        self.current_price = initial_price
        
    def reset(self):
        self.current_price = self.initial_price
        return self._get_state()
    
    def _get_state(self):
        return np.array([
            self.current_price,
            self.volatility,
            self.mean_reversion,
            np.random.normal(0, 1)
        ])
    
    def step(self, action: float):
        price_change = (
            self.mean_reversion * (self.initial_price - self.current_price) +
            self.volatility * np.random.normal(0, 1)
        )
        self.current_price += price_change
        
        shares_to_execute = int(action * self.total_shares / self.num_steps)
        shares_to_execute = max(0, min(shares_to_execute, self.total_shares))
        
        execution_price = self.current_price * (1 + 0.0001 * np.sqrt(shares_to_execute / 1000))
        
        reward = -abs(execution_price - self.current_price) * shares_to_execute
        
        return self._get_state(), reward, False


def train_td3_agent():
    """训练TD3智能体"""
    state_dim = 4
    action_dim = 1
    
    agent = TD3Agent(state_dim, action_dim)
    
    simulator = OrderExecutionSimulator()
    
    max_episodes = 500
    max_steps = 100
    batch_size = 256
    
    episode_rewards = []
    
    for episode in range(max_episodes):
        state = simulator.reset()
        episode_reward = 0
        
        for step in range(max_steps):
            action = agent.select_action(state, noise_scale=0.1)
            
            next_state, reward, done = simulator.step(action[0])
            
            agent.replay_buffer.push(state, action, reward, next_state, done)
            
            agent.train_step(batch_size)
            
            state = next_state
            episode_reward += reward
            
            if done or step == max_steps - 1:
                break
        
        episode_rewards.append(episode_reward)
        
        if episode % 50 == 0:
            avg_reward = np.mean(episode_rewards[-50:])
            print(f"Episode {episode}: Avg Reward = {avg_reward:.2f}")
    
    return agent, episode_rewards


if __name__ == "__main__":
    print("=" * 60)
    print("TD3强化学习订单执行策略")
    print("=" * 60)
    
    agent, rewards = train_td3_agent()
    
    print("\n训练完成，开始测试...")
    
    simulator = OrderExecutionSimulator(initial_price=100.0, volatility=0.025)
    state = simulator.reset()
    
    total_executed = 0
    total_cost = 0
    
    for step in range(100):
        action = agent.select_action(state, noise_scale=0.0)
        
        shares_to_execute = int(action[0] * simulator.total_shares / simulator.num_steps)
        shares_to_execute = max(0, min(shares_to_execute, simulator.total_shares - total_executed))
        
        execution_price = simulator.current_price * (1 + 0.0001 * np.sqrt(shares_to_execute / 1000))
        cost = abs(execution_price - simulator.current_price) * shares_to_execute
        
        total_executed += shares_to_execute
        total_cost += cost
        
        next_state, _, done = simulator.step(action[0])
        state = next_state
        
        if total_executed >= simulator.total_shares or done:
            break
        
        if step % 20 == 0:
            print(f"Step {step}: Executed {total_executed}/{simulator.total_shares} shares, Cost: ¥{total_cost:.2f}")
    
    print(f"\n最终统计:")
    print(f"  总执行量: {total_executed} / {simulator.total_shares}")
    print(f"  总交易成本: ¥{total_cost:.2f}")
    print(f"  成本率: {total_cost / (simulator.initial_price * simulator.total_shares) * 100:.4f}%")
```

### 数据来源
- Reddit: r/quant, r/algorithmictrading
- 小红书: 量化投资
- 知乎: 量化交易
